---
title: "Microbiome Data Simulation"
author: "Dahun Seo"
date: "2024-09-06"
geometry: "top=2cm, bottom=2cm, left=2.5cm, right=2.5cm"
header-includes:
- \usepackage[hangul]{kotex}
- \setmainhangulfont{나눔고딕}
- \setmainfont{Microsoft Sans Serif}
output:
  pdf_document: 
    latex_engine: xelatex
    highlight: zenburn
    toc: true
    toc_depth: 2
    number_sections: true
  html_document: default
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, collapse = T, comment = "#>")
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 50), tidy = TRUE)
options(tibble.print_min = 4L, tibble.print_max = 4L)
```

```{r message=FALSE, results="hide", include=FALSE}
rm(list = ls())
```

```{r message=FALSE, include=FALSE}
suppressPackageStartupMessages({
    library(MASS)
    library(dirmult)
    library(ape)
})
```

# Goal of proposed method

Goal

- The proposed methodology aims to identify features with significant signals for the outcome through an outcome model, while minimizing false discoveries.
- It focuses on variable selection utilizing phylogenetic tree information.
- Additionally, the method emphasizes obtaining valid confidence intervals for the estimator, laying the foundation for mediation analysis.

Main references

- Constructing predictive microbial signatures at multiple taxonomic levels, JASA (2017)
- Exact post-selection inference for the generalized lasso path, EJS (2018)

\clearpage

# Simulation Study

## The logistic normal (LN) distribution

- generate random binary tree with $p$ variables
- calculate cophenetic distanace between variables
- define variance-covariance matrix using distance matrix
\[
    \Sigma_{ij} = \exp(-d_{ij}) / 2
\]
- generate data from multivariate normal distribution modelled by covairates
\[
    M_i \sim \mathcal{N}_p(\mu_i, \Sigma)
\]
where
\[
    \mu_i = \alpha_0 + \alpha_t \text{treatment}_i + \alpha_\text{Sex} \text{Sex}_i + \alpha_\text{Age} \text{Age}_i
\]
- transformation to compositional data
\[
    Z_{ij} = \frac{\exp(M_{ij})}{\sum_{j=1}^p \exp(M_{ij})} 
\]
- log transformation or not
\[
    Z_{ij} = \log (Z_{ij})
\]
- generate outcome variable
\[
    Y_i = \beta_0 + \beta_t \text{treatment}_i + \beta_\text{Sex} \text{Sex}_i + \beta_\text{Age} \text{Age}_i + Z_i^\intercal \beta + \varepsilon
\]
where $\varepsilon \sim \mathcal{N}(0, \sigma^2)$.

```{r}
# The logistic normal (LN) distribution
n <- 20            # n: sample size
p <- 7             # p: number of features
noise_sigma <- 1    # noise_sigma: noise level for response

# covariate generation
set.seed(1234)
sex <- sample(c(0, 1), n, replace = TRUE)  # sex: 0 or 1
age <- rnorm(n, mean = 50, sd = 10)        # age: N(50, 10^2)

# treatment generation
treatment <- sample(c(0, 1), n, replace = TRUE) # treatment: 0 or 1

# parameters for normal distribution
base_mu <- rep(0, p)

# Create a random binary tree for the p features
random_tree <- ape::rcoal(p)
random_tree$tip.label <- paste0("x", 1:p)
```

```{r fig.align="center", out.width="60%"}
ape::plot.phylo(random_tree, type = "fan", main = paste("Random Binary Tree with", p, "Variables"))
```

```{r}
# Compute the distance matrix using cophenetic distances
dist_matrix <- stats::cophenetic(random_tree)

# variance-covariance setting
sigma <- exp(-dist_matrix) / 2

# sigma <- matrix(0, p, p)
# gamma <- 0.5
# for(i in 1:nrow(sigma)){
#     for(j in 1:nrow(sigma)){
#         sigma[i,j] <- gamma^(abs(i-j))   
#     }
# }

# coefficients
# alpha_sex <- rnorm(p, mean = 0.2, sd = 0.1)
# alpha_age <- rnorm(p, mean = 0.01, sd = 0.005)
# alpha_treatment <- rnorm(p, mean = 0.2, sd = 0.1)

# coefficients
alpha_sex <- 0
alpha_age <- 0
alpha_treatment <- 0

# calculate mu for each i
mu_matrix <- matrix(NA, n, p)
for (i in 1:n) {
    mu_matrix[i, ] <- base_mu + alpha_treatment * treatment[i] + alpha_sex * sex[i] + alpha_age * age[i]
}

# generation of logistic normal samples
z <- matrix(NA, n, p)
for (i in 1:n) {
    normal_sample <- MASS::mvrnorm(1, mu_matrix[i, ], sigma)
    exp_sample <- exp(normal_sample)
    z[i, ] <- exp_sample / sum(exp_sample)
}

# label
colnames(z) <- paste0("x", 1:p)
```

```{r fig.align="center", out.width="60%"}
head(z)
apply(z, 1, sum)
```

```{r}
# log transformation
z <- log(z)

# coefficients
beta_non_zero <-
    c(
        -3, 3, 2.5, -1, -1.5,
        3, 3, -2, -2, -2,
        1, -1, 3, -2, -1
    )

if (length(beta_non_zero) >= p) {
    beta <- beta_non_zero[1:p]
} else {
    beta <- c(beta_non_zero, rep(0, p - length(beta_non_zero)))
}

beta_sex <- 1
beta_age <- 1
beta_treatment <- 1
base_y <- rep(1, n)

y <- base_y + beta_treatment * treatment + beta_sex * sex + beta_age * age + as.vector(z %*% beta) + stats::rnorm(n, 0, sd = noise_sigma)
```

\clearpage

## The Dirichlet-multinomial (DM) distribution

- generate random binary tree with $p$ variables
- calculate cophenetic distanace between variables
- define variance-covariance matrix using distance matrix
\[
    \Sigma_{ij} = \exp(-d_{ij}) / 2
\]
- generate a sample $x_0$ from multivariate normal distribution $\mathcal{N}_p(0, \Sigma)$.
- generate samples $t_i$ from dirichlet distribution $\text{Dirichlet}(\alpha_i)$ where
$\alpha_i = x_0 \times \exp(\alpha_t \text{treatment}_i + \alpha_\text{Sex} \text{Sex}_i + \alpha_\text{Age} \text{Age}_i)$
- generate sequencing depth $n_i$ from negative biomial distribution
- generage count data from multinomial distribution
\[
    M_{i} \sim \text{Multi}(n_i, t_i)
\]
- transformation to compositional data
\[
    Z_{ij} = \frac{M_{ij}}{\sum_{j=1}^p M_{ij}} 
\]
- add the psudeo-count $0.5$ or not
- log transformation or not
\[
    Z_{ij} = \log (Z_{ij})
\]
- generate outcome variable
\[
    Y_i = \beta_0 + \beta_t \text{treatment}_i + \beta_\text{Sex} \text{Sex}_i + \beta_\text{Age} \text{Age}_i + Z_i^\intercal \beta + \varepsilon
\]
where $\varepsilon \sim \mathcal{N}(0, \sigma^2)$.

```{r}
# The Dirichlet-multinomial (DM) distribution
n <- 20            # n: sample size
p <- 7             # p: number of features
noise_sigma <- 1    # noise_sigma: noise level for response
depth <- stats::rnbinom(n, mu = 10000, size = 25)

# covariate generation
set.seed(1234)
sex <- sample(c(0, 1), n, replace = TRUE)  # sex: 0 or 1
age <- rnorm(n, mean = 50, sd = 10)        # age: N(50, 10^2)

# treatment generation
treatment <- sample(c(0, 1), n, replace = TRUE) # treatment: 0 or 1

# Create a random binary tree for the p features
random_tree <- ape::rcoal(p)
random_tree$tip.label <- paste0("x", 1:p)

# Compute the distance matrix using cophenetic distances
dist_matrix <- stats::cophenetic(random_tree)

# variance-covariance setting
cov_matrix <- exp(-dist_matrix) / 2

# coefficients
# base_alpha <- rep(1, p)
# alpha_sex <- rnorm(p, mean = 0.2, sd = 0.1)
# alpha_age <- rnorm(p, mean = 0.01, sd = 0.005)
# alpha_treatment <- rnorm(p, mean = 0.2, sd = 0.1)

# coefficients
# base_alpha <- rep(1, p)
# alpha_sex <- 0
# alpha_age <- 0
# alpha_treatment <- 0

# coefficients
base_alpha <- exp(MASS::mvrnorm(1, mu = rep(0, p), Sigma = cov_matrix))
alpha_sex <- 0
alpha_age <- 0
alpha_treatment <- 0

# calculate mu for each i
alpha_matrix <- matrix(NA, n, p)
for (i in 1:n) {
    alpha_matrix[i, ] <- base_alpha * exp(alpha_treatment * treatment[i] + alpha_sex * sex[i] + alpha_age * age[i])
}

# generation of Dirichlet-multinomial samples
z <- matrix(NA, nrow = n, ncol = p)
for (i in 1:n) {
    dirichlet_sample <- dirmult::rdirichlet(1, alpha_matrix[i, ])
    z[i, ] <- rmultinom(1, size = depth[i], prob = dirichlet_sample)
}

# label
colnames(z) <- paste0("x", 1:p)

# add the psudeo-count 0.5
z <- ifelse(z == 0, 0.5, z)

# calculate proportion
z <- z / rowSums(z)
```

```{r}
# check
head(z)
apply(z, 1, sum)
```

```{r}
# log transformation
z <- log(z)

# coefficients
beta_non_zero <-
    c(
        -3, 3, 2.5, -1, -1.5,
        3, 3, -2, -2, -2,
        1, -1, 3, -2, -1
    )

if (length(beta_non_zero) >= p) {
    beta <- beta_non_zero[1:p]
} else {
    beta <- c(beta_non_zero, rep(0, p - length(beta_non_zero)))
}

beta_sex <- 1
beta_age <- 1
beta_treatment <- 1
base_y <- rep(1, n)

y <- base_y + beta_treatment * treatment + beta_sex * sex + beta_age * age + as.vector(z %*% beta) + stats::rnorm(n, 0, sd = noise_sigma)

```

\clearpage

# performance measures

## empirical FDR
\[
    \widehat{\text{FDR}} = \mathbb{E}_N \left[ \frac{\lvert \{ j: \beta_j = 0 \text{ and } j \in \widehat{S}\}}{\lvert \widehat{S} \rvert \vee 1} \right]
\]
where $\mathbb{E}_N$ denotes the empirical average over replicated simulation.

## empirical power
\[
    \widehat{\text{Power}} = \mathbb{E}_N \left[ \frac{\lvert \{ j: \beta_j \neq 0 \text{ and } j \in \widehat{S}\}}{\lvert S^* \rvert} \right]
\]
where $\mathbb{E}_N$ denotes the empirical average over replicated simulation.

## empirical coverage?

## model mse?

# Competitive method

- Compositional knockoff filter for high‐dimensional regression analysis of microbiome data, Biometrics (2021)